src/lm/diff.cpp

```diff
@@ -0,0 +1,255 @@
+
+    set_stems(*table);
+    set_function_words(*table);
+}
+
+std::vector<std::pair<sentence, double>>
+    diff::candidates(const sentence& sent, bool use_lm /* = false */)
+{
+    use_lm_ = use_lm;
+    using pair_t = std::pair<sentence, double>;
+    auto comp = [](const pair_t& a, const pair_t& b)
+    {
+        return a.second < b.second;
+    };
+    std::priority_queue<pair_t, std::vector<pair_t>, decltype(comp)> candidates{
+        comp};
+    add(candidates, sent);
+
+    seen_.clear();
+    step(sent, candidates, 0);
+
+    std::vector<pair_t> sorted;
+    while (!candidates.empty())
+    {
+        sorted.emplace_back(std::move(candidates.top()));
+        candidates.pop();
+    }
+    std::reverse(sorted.begin(), sorted.end());
+    return sorted;
+}
+
+template <class PQ>
+void diff::add(PQ& candidates, const sentence& sent)
+{
+    seen_.insert(sent.to_string());
+    auto score = lambda_ * lm_.perplexity_per_word(sent)
+                 + (1.0 - lambda_) * sent.average_weight();
+    candidates.emplace(sent, score);
+    if (candidates.size() > max_cand_size_)
+        candidates.pop();
+}
+
+uint64_t diff::least_likely_ngram(const sentence& sent) const
+{
+    double min_prob = 1;
+    uint64_t best_idx = 0;
+    for (uint64_t i = n_val_; i < sent.size(); ++i)
+    {
+        auto ngram = sent(i - n_val_, i);
+        auto prob = lm_.log_prob(ngram);
+        if (prob < min_prob)
+        {
+            min_prob = prob;
+            best_idx = i - 1;
+        }
+    }
+
+    return best_idx;
+}
+
+template <class PQ>
+void diff::lm_ops(const sentence& sent, PQ& candidates, uint64_t depth)
+{
+    auto best_idx = least_likely_ngram(sent);
+
+    for (uint64_t i = 0; i < n_val_ && i < best_idx; ++i)
+    {
+        insert(sent, best_idx - i, candidates, depth);
+        remove(sent, best_idx - i, candidates, depth);
+        substitute(sent, best_idx - i, candidates, depth);
+    }
+
+    if (lm_generate_)
+    {
+        auto best = sent(best_idx - n_val_, best_idx);
+        best.pop_back();
+        try
+        {
+            for (auto& next : lm_.top_k(best, 3))
+            {
+                if (next.first == "</s>")
+                    continue;
+
+                sentence ins_cpy{sent};
+                ins_cpy.insert(best_idx, next.first,
+                               base_penalty_ + insert_penalty_);
+
+                if (seen_.find(ins_cpy.to_string()) == seen_.end())
+                {
+                    add(candidates, ins_cpy);
+                    step(ins_cpy, candidates, depth + 1);
+                }
+
+                sentence sub_cpy{sent};
+                sub_cpy.substitute(best_idx, next.first,
+                                   base_penalty_ + substitute_penalty_);
+
+                if (seen_.find(sub_cpy.to_string()) == seen_.end())
+                {
+                    add(candidates, sub_cpy);
+                    step(sub_cpy, candidates, depth + 1);
+                }
+            }
```

Comment:
Curious: have you profiled this part? Wondering if all these sentence copies are expensive. Is it worth having a more optimized `sentence` data structure that lets you model these edits better without the copy?